src/main/java/com/limechain/rpc/methods/state/StateRPCImpl.java

```diff
@@ -0,0 +1,215 @@
+        List<byte[]> keys = trieStorage.getKeysWithPrefixPaged(blockHash, prefix, startKey, limit);
+
+        String[][] result = new String[keys.size()][];
+        for (int i = 0; i < keys.size(); i++) {
+            result[i] = new String[]{HexUtils.toHexString(keys.get(i))};
+        }
+
+        return result;
+    }
+
+    public String stateGetStorage(final String keyHex, final String blockHashHex) {
+        if (!this.blockState.isInitialized()) {
+            return null;
+        }
+
+        byte[] key = StringUtils.hexToBytes(keyHex);
+        final Hash256 blockHash =
+                blockHashHex != null ? Hash256.from(blockHashHex) : blockState.getHighestFinalizedHash();
+
+        return trieStorage.getByKeyFromBlock(blockHash, new String(key))
+                .map(NodeData::getValue)
+                .map(HexUtils::toHexString)
+                .orElse(null);
+    }
+
+    public String stateGetStorageHash(final String keyHex, final String blockHashHex) {
+        if (!this.blockState.isInitialized()) {
+            return null;
+        }
+
+        byte[] key = StringUtils.hexToBytes(keyHex);
+        final Hash256 blockHash =
+                blockHashHex != null ? Hash256.from(blockHashHex) : blockState.getHighestFinalizedHash();
+
+        return trieStorage.getByKeyFromBlock(blockHash, new String(key))
+                .map(NodeData::getMerkleValue)
+                .map(HexUtils::toHexString)
+                .orElse(null);
+    }
+
+    public String stateGetStorageSize(final String keyHex, final String blockHashHex) {
+        if (!this.blockState.isInitialized()) {
+            return null;
+        }
+
+        byte[] key = StringUtils.hexToBytes(keyHex);
+        final Hash256 blockHash =
+                blockHashHex != null ? Hash256.from(blockHashHex) : blockState.getHighestFinalizedHash();
+
+        return trieStorage
+                .getByKeyFromBlock(blockHash, new String(key))
+                .map(NodeData::getValue)
+                .map(Array::getLength)
+                .map(String::valueOf)
+                .orElse(null);
+    }
+
+    public String stateGetMetadata(final String blockHashHex) {
+        if (!this.blockState.isInitialized()) {
+            return null;
+        }
+
+        final Hash256 blockHash =
+                blockHashHex != null ? Hash256.from(blockHashHex) : blockState.getHighestFinalizedHash();
+
+        final Runtime runtime = blockState.getRuntime(blockHash);
+        byte[] metadataBytes = runtime.call("Metadata_metadata");
+
+        return HexUtils.toHexString(metadataBytes);
+    }
+
+    public String stateGetRuntimeVersion(final String blockHashHex) {
+        if (!this.blockState.isInitialized()) {
+            return null;
+        }
+
+        final Hash256 blockHash =
+                blockHashHex != null ? Hash256.from(blockHashHex) : blockState.getHighestFinalizedHash();
```

Comment:
You can extract the initialized check and the getting of the `blockHash` for the nullable functions in a private method as it's done in all of them.

Reply:
or rather as separate functions as the hash is used everywhere